File: common/src/main/java/generations/gg/generations/core/generationscore/world/item/CustomTechnicalMachineItem.java
package generations.gg.generations.core.generationscore.world.item;

import net.minecraft.network.chat.Component;
import net.minecraft.world.item.ItemStack;

public class CustomTechnicalMachineItem extends MoveTeachingItem {
    public CustomTechnicalMachineItem(Properties properties) {
        super(properties);
    }

    @Override
    protected String getMoveString(ItemStack itemStack) {
        var tag = itemStack.getTag();

        if (tag != null) {
            if(tag.contains("move")) {
                return tag.getString("move");
            }
        }

        return null;
    }

    @Override
    public boolean isConsumed() {
        return false;
    }

    @Override
    public Component getName(net.minecraft.world.item.ItemStack stack) {
        var move = getMove(stack);
        return Component.literal("TM" + getNumber(stack) + " " + (move != null ? move.getDisplayName().toString() : "Blank"));
    }

    private String getNumber(ItemStack itemStack) {
        var tag = itemStack.getTag();

        return formatWithZeros(tag != null && tag.contains("number") ? tag.getInt("number") : 0);
    }

    private String formatWithZeros(int number) {
        var formattedNumber = Integer.toString(number);
        var zerosNeeded = 3 - formattedNumber.length();

        return zerosNeeded > 0 ? "0".repeat(zerosNeeded) + formattedNumber : formattedNumber.substring(0, 3);
    }

    public ItemStack createTm(int number, String move) {
        var stack = GenerationsItems.CUSTOM_TM.get().getDefaultInstance();
        var tag = stack.getOrCreateTag();
        tag.putInt("number", number);
        tag.putString("move", move);
        return stack;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/item/CustomTechnicalMachineItem.kt
package generations.gg.generations.core.generationscore.common.world.item

import generations.gg.generations.core.generationscore.common.util.extensions.asValue
import generations.gg.generations.core.generationscore.common.world.item.components.GenerationsDataComponents
import generations.gg.generations.core.generationscore.common.world.item.components.TmDetails
import net.minecraft.network.chat.Component
import net.minecraft.world.item.ItemStack

class CustomTechnicalMachineItem(properties: Properties) : MoveTeachingItem(properties) {
    override fun getMoveString(itemStack: ItemStack): String? {
        val tag = itemStack.get(GenerationsDataComponents.TM_DETAILS.asValue<TmDetails>())
        if (tag != null) {
            return tag.move
        }
        return null
    }

    override val isConsumed: Boolean
        get() = false

    override fun getName(stack: ItemStack): Component {
        val move = getMove(stack)
        return Component.literal("TM" + getNumber(stack) + " " + (move?.displayName?.toString() ?: "Blank"))
    }

    private fun getNumber(itemStack: ItemStack): String {
        val tag = itemStack.get(GenerationsDataComponents.TM_DETAILS.asValue<TmDetails>())
        return (tag?.number ?: 0).let(::formatWithZeros)
    }

    private fun formatWithZeros(number: Int): String {
        val formattedNumber = Integer.toString(number)
        val zerosNeeded = 3 - formattedNumber.length
        return if (zerosNeeded > 0) "0".repeat(zerosNeeded) + formattedNumber else formattedNumber.substring(0, 3)
    }

    fun createTm(move: String, number: Int): ItemStack {
        return GenerationsItems.CUSTOM_TM.value().defaultInstance.also { it.set(GenerationsDataComponents.TM_DETAILS.asValue(), TmDetails(move, number)) }
    }
}
